### ISSUE TITLE
Allow rendering deterministic sites

### ISSUE THREAD

**Issue by fehiepsi** [pyro-ppl/numpyro#1383]
Currently, we skip deterministic sites while rendering the model. It would be nice to support plotting deterministic nodes (e.g. to make the rendered model clearer when using reparam handlers). This is especially useful for hierarchical models like the one in #1295

**Comment by YanniPapandreou** [comment]
Hi, is anyone working on this? If not I'd be happy to give this a try - I'm quite new to this so any guidance would be great :)

**Comment by fehiepsi** [comment]
Thanks,  @YanniPapandreou! The inspect logic is in [this file](https://github.com/pyro-ppl/numpyro/blob/f81ce325de402d9a9891497136a36e3b44f982cc/numpyro/infer/inspect.py#L192). Currently, we only consider messages with `sample` type. You can modify the dictionaries in that method to also include `deterministic` sites. You can get started by making a model with deterministic site, obtain a trace of it, play around with provenance logic.

When you start looking into provenance, you will see that we use log_prob method to determine the dependency. For deterministic nodes, you can simply use `identity` function instead.

**Comment by YanniPapandreou** [comment]
Thanks @fehiepsi ! I have started looking into this. 

Do you have any resources for learning about provenance logic? I have managed to get the deterministic sites but I am not sure I understand how `log_prob` is used for samples to get the dependency. I am also not sure about which `identity` function you are referring to is this **the** identity function which seems to be implemented in `numpyro.utils`?

**Comment by fehiepsi** [comment]
Sorry @YanniPapandreou, I miss your response! Let me make the previous comment clearer.

#### provenance

This is used to see which variable in the output depends on which variable in the input. For example,
```python
from numpyro.ops.provenance import ProvenanceArray, eval_provenance, get_provenance
a = ProvenanceArray(1., frozenset({"a"}))
b = ProvenanceArray(2., frozenset({"b"}))
c = 3.
f = lambda a, b, c: (a + b + c, a - c)
d, e = eval_provenance(f, a, b, c)
assert get_provenance(d) == frozenset({"a", "b"})
assert get_provenance(e) == frozenset({"a"})
```

#### model dependencies

Given a model with latent variables `a, b, c`. We will convert those variables to a dictionary of [ProvenanceArray](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/inspect.py#L201)s (you can add deterministic sites into this dictionary. If `c` depends on `a` and `b`, provenance of its `log_prob` will be `a` and `b`. We evaluate log_prob of each site in the following function

```python
def _get_log_probs(model, model_args, model_kwargs, sample):
    # Note: We use seed 0 for parameter initialization.
    with handlers.trace() as tr, handlers.seed(rng_seed=0), handlers.substitute(
        data=sample
    ):
        model(*model_args, **model_kwargs)
    return {
        name: site["fn"].log_prob(site["value"])
        for name, site in tr.items()
        if site["type"] == "sample"
    }
```

For include deterministic sites, you will need to add `"deterministic"` to [this list](https://github.com/pyro-ppl/numpyro/blob/b22d9c1ccbfa543852bfa2c0d63d206d0b47ac92/numpyro/handlers.py#L777) and replace the return statement above by
```python
    provenance_arrays = {}
    for name, site in tr.items():
        if site["type"] == "sample":
            provenance_arrays[name] = site["fn"].log_prob(site["value"])
        elif site["type"] == "deterministic":
            provenance_arrays[name] = site["value"]  # this is identity function I mentioned above
    return provenance_arrays
```

[Those lines](https://github.com/pyro-ppl/numpyro/blob/b22d9c1ccbfa543852bfa2c0d63d206d0b47ac92/numpyro/infer/inspect.py#L200-L210) implement the provenance logic. The output `sample_deps` will include all dependencies that you need. :)

Please let me know if something is not clear.

**Comment by YanniPapandreou** [comment]
@fehiepsi thanks for sharing! I will try to take a look at this when I get a chance :)

**Comment by YanniPapandreou** [comment]
Hi @fehiepsi, I have started working on this. I have made some progress but I am still a bit stuck.

For the following model:

```python
def model(data):
    m = numpyro.sample("m", dist.Normal(0, 1))
    sd = numpyro.sample("sd", dist.LogNormal(m, 1))
    m_transformed = numpyro.deterministic("m_transformed", m + 1)
    with numpyro.plate("N", len(data)):
        numpyro.sample("obs", dist.Normal(m_transformed, sd), obs=data)
```

I was able to obtain this diagram:

![image](https://user-images.githubusercontent.com/15948162/205464703-108f544f-2a88-4e61-8d84-f52cb42b742c.png)

I tried adjusting the code to get `obs` to depend on `m_transformed` but then `m_transformed` no longer depends on `m` as I detail below. I think I am missing something. 

I largely followed what you suggested, though I changed `get_model_relations`  (and `get_log_probs`) [here](https://github.com/pyro-ppl/numpyro/blob/b22d9c1ccbfa543852bfa2c0d63d206d0b47ac92/numpyro/infer/inspect.py#L256) instead of `get_dependencies` and `_get_log_probs` which you referred to above. Please let me know if I am wrong but it seems to me that `get_model_relations` is used in `render_model` instead of `get_dependencies`.  

I have added `"deterministic"` to [the list](https://github.com/pyro-ppl/numpyro/blob/b22d9c1ccbfa543852bfa2c0d63d206d0b47ac92/numpyro/handlers.py#L777) you suggested as detailed in this commit in a fork I made [here](https://github.com/pyro-ppl/numpyro/commit/76f35f1d7d51107e99a015ffae9486e38a3cb287).

The changes I made are in this other commit [here](https://github.com/pyro-ppl/numpyro/commit/bbc4769bb12e0e17f770929fa78028019921fa0a).

If I uncomment what I added at [line 386](https://github.com/pyro-ppl/numpyro/commit/bbc4769bb12e0e17f770929fa78028019921fa0a) I instead get this diagram:

![image](https://user-images.githubusercontent.com/15948162/205465246-cfcff80b-c754-4128-ab1b-2a67e1a6ce0d.png)

I think that uncommenting the line is wrong as I am basically creating a new ProvenanceArray for `m_transformed` which only depends on itself.

Do you have any suggestions on how I can make `obs` realise it depends on `m_transformed`?

**Comment by fehiepsi** [comment]
Hi @YanniPapandreou, could you create a pull request? It is easier for me to comment on the top of the changes. Thanks!

**Comment by YanniPapandreou** [comment]
> Hi @YanniPapandreou, could you create a pull request? It is easier for me to comment on the top of the changes. Thanks!

Sure, this has been created: #1508. Please let me know if anything isn't clear or is wrong (this is my first time creating a pull request! :smile: )

**Comment by fehiepsi** [comment]
Resolved in #1508. Thanks @YanniPapandreou!
